src/lib/utils/enum_constant.hpp

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <boost/hana/hash.hpp>
+#include <boost/hana/value.hpp>
+
+#include <type_traits>
+
+namespace opossum {
+
+/**
+ * @defgroup Definition of Enum Constants
+ *
+ * An enum constant extends the concept of an integral constant to enum types.
+ * This makes it possible to use enums in hana data structures such as maps.
```

Comment:
Could you give an example of what is made possible with this? If it is easier, maybe even one from outside of Hyrise?

Reply:
An enum constant is the same as a type constant (created through `hana::type_c<int>`) or an integral constant (`std::integral_constant<int, 4>`). They make it possible to create objects that store values available at compile time.

Right now it’s used to map `EncodingType`s to supperted data types as well as to encoded column classes. Both mappings are used to instantiate templates in various situations. So for example the map from encoding type to supported data types ensures that the template method in column encoders is instantiated with all supported types. The other map is needed to resolve the type of encoded columns.

A general example of where they are applied is called tag dispatching. Tag dispatching means overloading method or functions not only for various types but also for individual values. Let’s say we decide to have several data types with the same underlying type. For example we could add the new type `Date` which is just an `int32_t`.

```c++
using int32_t = int;
class Date;

enum class DataType { Int, Date };
auto data_types = hana::make_tuple(
    hana::make_pair(enum_c<DataType::Int>, type_c<int32_t>),
    hana::make_pair(enum_c<DataType::Date>, type_c<Date>));
```

This is how our definition could look like. At the moment, we have specialised implementations for data types in several classes (e.g. `AggregateTraits`). We don’t use the enum there to decide what to do, but the actual types. This wouldn’t be possible anymore. What we could do instead, is use the enums using tag-dispatching. We could update `resolve_data_type` to output `enum_constant`s instead of `hana::basic_type`s and then do the following:

```c++
auto column = std::shared_ptr<BaseValueColumn>{ ... };

resolve_data_type(data_type, [&](auto data_type_c) {
  process(data_type_c, column);
});


void process(enum_constant<DataType::Int>, std::shared_ptr<BaseValueColumn> base_column) {
  auto column = std::static_pointer_cast<ValueColumn<int32_t>>(base_column);

  // process column as integer
}

void process(enum_constant<DataType::Date>, std::shared_ptr<BaseValueColumn> base_column) {
  auto column = std::static_pointer_cast<ValueColumn<int32_t>>(base_column);

  // process column as date
}
```

Reply:
Got it, thanks for the explanation. Could you move that into a comment?

Reply:
I’ve added a wiki entry and a link in the code.